src/common/WeightedPriorityQueue.h

```diff
@@ -0,0 +1,369 @@
+    }
+  }
+
+  void remove_by_class(K k, std::list<T> *out = 0) override final {
+    for (typename SubQueues::iterator i = queue.begin();
+	 i != queue.end(); ++i) {
+      wrr_size -= i->second.remove_by_class(k, out);
+      unsigned priority = i->first;
+      if (i->second.empty()) {
+	remove_queue(priority);
+      }
+    }
+    for (typename SubQueues::iterator i = high_queue.begin();
+	 i != high_queue.end();
+	 ) {
+      high_size -= i->second.remove_by_class(k, out);
+      if (i->second.empty()) {
+	high_queue.erase(i++);
+      } else {
+	++i;
+      }
+    }
+  }
+
+  void enqueue_strict(K cl, unsigned priority, T item) override final {
+    high_queue[priority].enqueue(cl, 0, item);
+    ++high_size;
+  }
+
+  void enqueue_strict_front(K cl, unsigned priority, T item) override final {
+    high_queue[priority].enqueue_front(cl, 0, item);
+    ++high_size;
+  }
+
+  void enqueue(K cl, unsigned priority, unsigned cost, T item) override final {
+    if (cost > max_cost) {
+      max_cost = cost;
+    }
+    create_queue(priority)->enqueue(cl, cost, item);
+    ++wrr_size;
+  }
+
+  void enqueue_front(K cl, unsigned priority, unsigned cost, T item) override final {
+    if (cost > max_cost) {
+      max_cost = cost;
+    }
+    create_queue(priority)->enqueue_front(cl, cost, item);
+    ++wrr_size;
+  }
+
+  bool empty() const override final {
+    assert(total_priority >= 0);
+    assert((total_priority == 0) || !queue.empty());
+    return (high_size + wrr_size  == 0) ? true : false;
+  }
+
+  void inc_dq() {
+    ++dq;
+    if (dq == queue.rend()) {
+      dq = queue.rbegin();
+    }
+  }
+
+  T dequeue() override final {
+    assert(!empty());
+
+    if (!high_queue.empty()) {
+      T ret = high_queue.rbegin()->second.front().second;
+      high_queue.rbegin()->second.pop_front();
+      if (high_queue.rbegin()->second.empty()) {
+	high_queue.erase(high_queue.rbegin()->first);
+      }
+      --high_size;
+      return ret;
+    }
+    while (true) {
+      // If there is only one priority, run it.
+      if (dq->second.size() == wrr_size) {
+	break;
+      } else {
+	// Pick a new priority out of the total priority.
+	unsigned prio = rand() % total_priority;
+	typename SubQueues::iterator i = queue.begin();
+	unsigned tp = i->first;
+	// Find the priority coresponding to the picked number.
+	// Add low priorities to high priorities until the picked number
+	// is less than the total and try to dequeue that priority.
+	while (prio > tp) {
+	  ++i;
+	  tp += i->first;
+	}
+	dq = (typename SubQueues::reverse_iterator) ++i;
+	// Flip a coin to see if this priority gets to run based on cost.
+	// The next op's cost is multiplied by .9 and subtracted from the
+	// max cost seen. Ops with lower costs will have a larger value
+	// and allow them to be selected easier than ops with high costs.
+	if (max_cost == 0 || rand() % max_cost <=
+	    (max_cost - ((dq->second.front().first * 9) / 10))){
+	  break;
+	}
+      }
+      inc_dq();
```

Comment:
That is a relic from the round robin code I tried before. It doesn't do anything other than burn CPU now. dq is reassigned in each new pass so it doesn't matter what it is at this point.